pennylane/plugins/expt_tensornet_tf.py

```diff
@@ -0,0 +1,360 @@
+    r"""Two-qubit controlled rotation about the y axis.
+
+    Args:
+        theta (float): rotation angle
+    Returns:
+        array: unitary 4x4 rotation matrix :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R_y(\theta)`
+    """
+    theta = tf.cast(theta, dtype=tf.complex128)
+    return tf.cos(theta/4)**2 * II - 1j*tf.sin(theta/2)/2 * IY + tf.sin(theta/4)**2 * ZI + 1j*tf.sin(theta/2)/2 * ZY
+
+
+def CRotz(theta):
+    r"""Two-qubit controlled rotation about the z axis.
+
+    Args:
+        theta (float): rotation angle
+    Returns:
+        array: unitary 4x4 rotation matrix
+        :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R_z(\theta)`
+    """
+    theta = tf.cast(theta, dtype=tf.complex128)
+    return tf.cos(theta/4)**2 * II - 1j*tf.sin(theta/2)/2 * IZ + tf.sin(theta/4)**2 * ZI + 1j*tf.sin(theta/2)/2 * ZZ
+
+
+def CRot3(a, b, c):
+    r"""Arbitrary two-qubit controlled rotation using three Euler angles.
+
+    Args:
+        a,b,c (float): rotation angles
+    Returns:
+        array: unitary 4x4 rotation matrix
+        :math:`|0\rangle\langle 0|\otimes \mathbb{I}+|1\rangle\langle 1|\otimes R(a,b,c)`
+    """
+    return CRotz(c) @ (CRoty(b) @ CRotz(a))
+
+
+class TensorNetworkTF(TensorNetwork):
+    """Experimental TensorFlow Tensor Network simulator device for PennyLane.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+    """
+    name = "PennyLane TensorNetwork (TensorFlow) simulator plugin"
+    short_name = "expt.tensornet.tf"
+    _capabilities = {"model": "qubit", "tensor_observables": True, "provides_jacobian": True}
+
+    _operation_map = {
+        "QubitStateVector": None,
+        "BasisState": None,
+        "QubitUnitary": unitary,
+        "PauliX": X,
+        "PauliY": Y,
+        "PauliZ": Z,
+        "Hadamard": H,
+        "S": S,
+        "T": T,
+        "CNOT": CNOT,
+        "SWAP": SWAP,
+        "CSWAP": CSWAP,
+        "CZ": CZ,
+        "PhaseShift": Rphi,
+        "RX": Rotx,
+        "RY": Roty,
+        "RZ": Rotz,
+        "Rot": Rot3,
+        "CRX": CRotx,
+        "CRY": CRoty,
+        "CRZ": CRotz,
+        "CRot": CRot3,
+    }
+
+    def __init__(self, wires, shots=1000, analytic=True):
+        super().__init__(wires, shots)
+        self.backend = "tensorflow"
+
+        self.eng = None
+        self.analytic = True
+        self._nodes = []
+        self._edges = []
+        zero_state = np.zeros([2] * wires)
+        zero_state[tuple([0] * wires)] = 1.0
+        self._zero_state = tf.constant(zero_state, dtype=tf.complex128)
+        # TODO: since this state is separable, can be more intelligent about not making a dense matrix
+        self._state_node = self._add_node(
+            self._zero_state, wires=tuple(w for w in range(wires)), name="AllZeroState"
+        )
+        self._free_edges = self._state_node.edges[:]  # we need this list to be distinct from self._state_node.edges
```

Comment:
Unfortunately, there are a couple of lines in `TensorNet.__init.__()` which is incompatible with TensorFlow:

* Uses a NumPy array as the initial state tensor
* Sets the initial state via `self._zero_state[tuple([0] * wires)] = 1.0`. Array assignment is not supported by TensorFlow.

This is why the entire method was copied, instead of being called --- these two original lines cause exceptions when used with the TensorFlow backend. Is there a better solution?

Reply:
just make those two offending lines private methods, implemented separately in each class

Reply:
Done